### PR TITLE
Manage co2 emission

### PR DESCRIPTION
**Pull request by azime**
manage co2 emission

### CODE REVIEW

File: source/type/type.h [hunk 1]
```diff
@@ -548,7 +548,7 @@ struct CommercialMode : public Header, Nameable{
 
 struct PhysicalMode : public Header, Nameable{
     const static Type_e type = Type_e::PhysicalMode;
-    double co2_emission = 0.;
+    double co2_emission = -1.;
```
Comment (l-vincent-l) [inline source/type/type.h]: please use std::numeric_limits<double>::max()

File: source/ed/build_helper.cpp [hunk 2]
```diff
@@ -397,6 +397,27 @@ void builder::connection(const std::string & name1, const std::string & name2, f
...
+         return 0.;
+     }
+     if (uri == "Bus"){
+         return 132.;
+     }
+     if (uri == "Car"){
+         return 184.;
+     }
+    return -1.;
+ }
```
Comment (l-vincent-l) [inline source/ed/build_helper.cpp:418]: You don't need this function, just set the value when you initialized commercial modes

File: source/ed/types.h [hunk 3]
```diff
@@ -187,7 +187,7 @@ struct CommercialMode : public Header, Nameable{
 
 struct PhysicalMode : public Header, Nameable{
     const static nt::Type_e type = nt::Type_e::PhysicalMode;
-    double co2_emission = 0.;
+    double co2_emission = -1.;
```
Comment (l-vincent-l) [inline source/ed/types.h]: I don't understand, in ed you store it in physical_mode and in commercial_mode in navitia.

Comment (azime) [inline source/ed/types.h]: physical_mode in ed and in navitia

File: source/ed/build_helper.cpp (hunk 2 above)
Comment (azime) [inline source/ed/build_helper.cpp:418]: co2_emission in physical_mode object

**Comment by azime** [comment]
corrections

File: source/jormungandr/jormungandr/interfaces/v1/fields.py [hunk 4]
```diff
@@ -344,6 +344,17 @@ def output(self, key, obj):
         return response
 
 
+class Co2Emission(fields.Raw):
+    def output(self, key, obj):
+        if not obj.HasField("co2_emission"):
+            logging.warn("trying to output wrongly formated object as geojson, we skip")
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py]: ?

**Comment by azime** [comment]
corrections

File: source/type/type.h [hunk 5]
```diff
@@ -548,7 +548,7 @@ struct CommercialMode : public Header, Nameable{
 
 struct PhysicalMode : public Header, Nameable{
     const static Type_e type = Type_e::PhysicalMode;
-    double co2_emission = 0.;
+    double co2_emission = invalid_idx;
```
Comment (kinnou02) [inline source/type/type.h]: why?

File: source/jormungandr/jormungandr/interfaces/v1/fields.py [hunk 6]
```diff
@@ -344,6 +344,17 @@ def output(self, key, obj):
         return response
 
 
+class Co2Emission(fields.Raw):
+    def output(self, key, obj):
+        if not obj.HasField("co2_emission"):
+            logging.warn("co2_emission not in object, we skip")
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py]: Each section without Co2 will generate a warning in the log I think

File: source/type/type.h (hunk 5 above)
Comment (azime) [inline source/type/type.h]: 2 values possibles for co2_emission:
XX : value defined
invalid_idx : if undefined co2_emission value 

Commented by l-vincent-l : please use std::numeric_limits::max()

File: source/jormungandr/jormungandr/interfaces/v1/fields.py (hunk 6 above)
Comment (azime) [inline source/jormungandr/jormungandr/interfaces/v1/fields.py]: OK

File: source/type/type.h (hunk 5 above)
Comment (kinnou02) [inline source/type/type.h]: invalid_idx is a uint_32_t, co2 emission is a double so the remark of vincent is valid, you have to use is directly

**Comment by azime** [comment]
corrections

File: source/ed/types.h (hunk 3 above)
Comment (kinnou02) [inline source/ed/types.h]: max double no?

File: source/type/pb_converter.cpp [hunk 7]
```diff
@@ -1615,14 +1617,22 @@ pbnavitia::StreetNetworkMode convert(const navitia::type::Mode_e& mode) {
 
 }
 
-void fill_co2_emission(pbnavitia::Section *pb_section, const type::VehicleJourney* vehicle_journey){
+void fill_co2_emission_by_mode(pbnavitia::Section *pb_section, const nt::Data& data, const std::string& mode_uri){
+    if (!mode_uri.empty()){
+      const auto it_physical_mode = data.pt_data->physical_modes_map.find(mode_uri);
+      if ((it_physical_mode != data.pt_data->physical_modes_map.end()) && (it_physical_mode->second->co2_emission != std::numeric_limits<double>::max())){
```
Comment (kinnou02) [inline source/type/pb_converter.cpp]: This line is way too long

File: source/routing/raptor_api.cpp [hunk 8]
```diff
@@ -117,7 +117,25 @@ static void fill_section(pbnavitia::Section *pb_section, const type::VehicleJour
         add_coord(stop_times.front()->journey_pattern_point->stop_point->coord, pb_section);
         add_coord(stop_times.back()->journey_pattern_point->stop_point->coord, pb_section);
     }
-    fill_co2_emission(pb_section, vj);
+    fill_co2_emission(pb_section, d, vj);
+}
+
+void co2_emission_aggregator(pbnavitia::Journey* pb_journey){
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: as this method is only used in this file, and not exported in the .h, please add "static" before:

``` c++
static void co2_emission_aggregator(pbnavitia::Journey* pb_journey){
```

File: source/routing/raptor_api.cpp [hunk 9]
```diff
@@ -472,8 +491,10 @@ static void add_pathes(EnhancedResponse& enhanced_response,
             auto destination_pb = pb_journey->mutable_sections(pb_journey->sections_size()-1)->mutable_destination();
             destination_pb->Clear();
             fill_pb_placemark(destination, d, destination_pb, 2);
+            co2_emission_aggregator(pb_journey);
         }
     }
+
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: useless line

File: source/ed/build_helper.cpp (hunk 2 above)
Comment (TeXitoi) [inline source/ed/build_helper.cpp:418]: you can static this method and remove the prototype from the header file.

**Comment by TeXitoi** [comment]
OK for me after the asked corrections.

**Comment by azime** [comment]
Corrections

File: source/ed/build_helper.cpp (hunk 2 above)
Comment (TeXitoi) [inline source/ed/build_helper.cpp:418]: You forgot this one: please make this method a static function (`static double get_co2_emission(const std::strings& uri){`, then...

File: source/ed/build_helper.h [hunk 10]
```diff
@@ -147,6 +147,7 @@ struct builder{
     void build_autocomplete();
 
     nt::MetaVehicleJourney* get_or_create_metavj(const std::string name);
+    double get_co2_emission(const std::string& uri);
```
Comment (TeXitoi) [inline source/ed/build_helper.h]: ... remove its prototype.

**Comment by azime** [comment]
- Corrections
-  tests added

**Comment by kinnou02** [comment]
Jenkins, test this please

**Comment by azime** [comment]
Use boost::optional<double> for co2_emission variable

**Comment by antoine-de** [comment]
Jenkins, test this please

**Comment by TeXitoi** [comment]
Jenkins, test this please

**Pull request merged**